### DISCUSSION TITLE
コンテキストスイッチ処理の詳細について

### DISCUSSION THREAD

**Discussion by keisuke713** [nuta/microkernel-book#14] [Help] [answered]
### 概要

[こちら](https://github.com/nuta/microkernel-book/blob/main/kernel/riscv32/switch.S#L8)の関数の [この辺り](https://github.com/nuta/microkernel-book/blob/main/kernel/riscv32/switch.S#L12-L24)の処理が今ひとつ理解できないのですが、やっていることとしては `レジスタに入っている値をプロセスの持つスタック内のメモリに退避する` で合っていますか？

**Comment by keisuke713** [comment]
また各レジスタ( `s0~s11` )には何が保存されている想定なのでしょうか？ローカル変数などですか？

**Comment by nuta** [comment, marked as answer]
> レジスタに入っている値をプロセスの持つスタック内のメモリに退避する で合っていますか

その通りです！C言語で考えると次のようなイメージです。

```c
struct frame {
    uint32_t ra;
    uint32_t s0;
    uint32_t s1;
    uint32_t s2;
    uint32_t s3;
    uint32_t s4;
    uint32_t s5;
    uint32_t s6;
    uint32_t s7;
    uint32_t s8;
    uint32_t s9;
    uint32_t s10;
    uint32_t s11;
};

// 擬似コード。ra, s0, ...はレジスタを操作しているつもり。
void riscv32_task_switch(uint32_t *prev_sp, uint32_t *next_sp) {
    struct frame *f_prev = (スタックポインタからframe構造体の大きさだけ引いたアドレス); // addi sp, sp, -13 * 4
    f_prev->ra = ra;         // sw ra,  0  * 4(sp)
    f_prev->s0 = s0;         // sw s0,  1  * 4(sp)
    f_prev->s1 = s1;         // sw s1,  2  * 4(sp)
    f_prev->s2 = s2;         // sw s2,  3  * 4(sp)
    f_prev->s3 = s3;         // sw s3,  4  * 4(sp)
    f_prev->s4 = s4;         // sw s4,  5  * 4(sp)
    f_prev->s5 = s5;         // sw s5,  6  * 4(sp)
    f_prev->s6 = s6;         // sw s6,  7  * 4(sp)
    f_prev->s7 = s7;         // sw s7,  8  * 4(sp)
    f_prev->s8 = s8;         // sw s8,  9  * 4(sp)
    f_prev->s9 = s9;         // sw s9,  10 * 4(sp)
    f_prev->s10 = s10;       // sw s10, 11 * 4(sp)
    f_prev->s11 = s11;       // sw s11, 12 * 4(sp)

    *prev_sp = sp;      // sw sp, (a0)  // 実行中タスクのスタックポインタを保存
    sp = *next_sp;      // lw sp, (a1)  // 次のタスクのスタックポインタを復元

    ra = f_next->ra; // lw ra,  0  * 4(sp)
    s0 = f_next->s0; // lw s0,  1  * 4(sp)
    s1 = f_next->s1; // lw s1,  2  * 4(sp)
    s2 = f_next->s2; // lw s2,  3  * 4(sp)
    s3 = f_next->s3; // lw s3,  4  * 4(sp)
    s4 = f_next->s4; // lw s4,  5  * 4(sp)
    s5 = f_next->s5; // lw s5,  6  * 4(sp)
    s6 = f_next->s6; // lw s6,  7  * 4(sp)
    s7 = f_next->s7; // lw s7,  8  * 4(sp)
    s8 = f_next->s8; // lw s8,  9  * 4(sp)
    s9 = f_next->s9; // lw s9,  10 * 4(sp)
    s10 = f_next->s10; // lw s10, 11 * 4(sp)
    s11 = f_next->s11; // lw s11, 12 * 4(sp)
    sp = (spにframe構造体の大きさだけ足したアドレス); // addi sp, sp, 13 * 4

    goto ra; // raに復元された値へジャンプ (return)
}
```

まず、スタック上で退避領域 (`frame` 構造体) を割り当てて、各フィールドにレジスタの値を次のように書き込んでいきます。

```
sw レジスタ名,  フィールドの位置 * 4(sp)
```

ここで、`4`は各フィールドのサイズ (4バイト) を指します。ベースアドレス (`sp`) から0バイト目、4バイト目、8バイト目、... と書き込んでいく形になります。

> また各レジスタ( s0~s11 )には何が保存されている想定なのでしょうか？ローカル変数などですか？

こちらもその通りです！CPUには「各レジスタをどのように利用すべきか」という規約 (呼出規約) があります ([こちらのZennの記事](https://zenn.dev/tetsu_koba/articles/b7b31b372f0f40)が参考になるかもしれません)。`s0-s11`その中では「呼び出し先保存レジスタ (callee-saved registers)」と呼ばれる類のもので、おっしゃる通りローカル変数を保持しています。

それらのレジスタについては、関数呼び出しの前後で同じ値を保持している (破壊される際はちゃんと呼び出し**先** が保存・復元する責任を持つ) ことを仮定した機械語をコンパイラが生成します。そこで、特別扱いして別途保存してあげるのが上記の処理になります。

**Reply by keisuke713** [reply]
サンプルコードまでありがとうございます！！！
疑問に思っていたことを理解する事ができました。

> s0-s11その中では「呼び出し先保存レジスタ (callee-saved registers)」と呼ばれる類のもので、おっしゃる通りローカル変数を保持しています。
それらのレジスタについては、関数呼び出しの前後で同じ値を保持している (破壊される際はちゃんと呼び出し先 が保存・復元する責任を持つ) ことを仮定した機械語をコンパイラが生成します。

なるほど、 `s0-s11` だけを保存していることも違和感があったのですが、呼び出し先である `riscv32_task_switch` が保存、復元する責務を負っているのですね。
逆に言えばその他のレジスタ( `a0-a7`など)などは呼び出し元でスタックなどに保存しておかなければならないということでしょうか？

**Reply by nuta** [reply]
その通りです！逆に呼び出し元保存レジスタは、名前の通り関数を呼び出す**前**に保存されるべきレジスタたちです。こちらについてはコンパイラが保存・復元する機械語を生成してくれるので、コンテキストスイッチ処理のコード上 (`riscv32_task_switch`) では登場しません。

**Reply by keisuke713** [reply]
> コンパイラが保存・復元する機械語を生成してくれるので、コンテキストスイッチ処理のコード上 (riscv32_task_switch) では登場しません。

なるほどです 🙏 

全然質問じゃないのですが、コンテキストスイッチの際に `ra` レジスタの内容も保存しているから切り替えた後に前回の続きから処理を進めることができるのですね。
